File: Desafio #01 - Numero palindromo/DART/numeroPalindromoTest.dart
import 'dart:math';

void main() {
  final number = generarPalindromoBigInt(10000);
  //print(number);
  // Medir el tiempo de ejecución de la función 1
  Stopwatch reloj = Stopwatch();

  reloj.start();
  final result1 = isPalindrome3(number);
  reloj.stop();
  print(result1);
  var tiempoFuncionUno = reloj.elapsedMicroseconds;

  // Medir el tiempo de ejecución de la función 2
  reloj.reset();
  reloj.start();
  final result2 = isPalindrome3(number);
  reloj.stop();
  print(result2);
  var tiempoFuncionDos = reloj.elapsedMicroseconds;
  reloj.reset();
  // Comparar los tiempos de ejecución
  print('Tiempo de ejecución de la Función 1: $tiempoFuncionUno us');
  print('Tiempo de ejecución de la Función 2: $tiempoFuncionDos us');

  if (tiempoFuncionUno < tiempoFuncionDos) {
    print('La Función 1 es más rápida.');
  } else if (tiempoFuncionUno > tiempoFuncionDos) {
    print('La Función 2 es más rápida.');
  } else {
    print('Ambas funciones tienen el mismo rendimiento.');
  }
}

bool isPalindrome(BigInt number) {
  String numberStr = number.toString();
  String reverseNumber = numberStr.split('').reversed.join('');

  return numberStr == reverseNumber;
}

bool isPalindrome2(BigInt number) {
  final numberStr = number.toString();
  final numberLength = numberStr.length;

  for (int i = 0; i < numberLength ~/ 2; i++) {
    if (numberStr[i] != numberStr[numberLength - i - 1]) {
      return false;
    }
  }
  return true;
}

bool isPalindrome3(BigInt number) {
  // Manejar el caso especial de 0 directamente
  if (number == BigInt.zero) {
    return true;
  }

  // Copiamos el número original para compararlo más tarde
  BigInt originalNumber = number;
  BigInt reversedNumber = BigInt.zero;

  // Invertimos el número
  while (number > BigInt.zero) {
    // Obtenemos el dígito actual
    BigInt digit = number % BigInt.from(10);
    
    // Añadimos el dígito al número invertido
    reversedNumber = reversedNumber * BigInt.from(10) + digit;

    // Eliminamos el último dígito del número original (lo desplazamos un lugar a la derecha)
    number = number ~/ BigInt.from(10);
  }

  // Comparamos el número original con su versión invertida
  return originalNumber == reversedNumber;
}

// Función para generar un BigInt de una longitud específica con dígitos aleatorios
BigInt generarBigIntAleatorio(int longitud) {
  Random random = Random();
  String numeroStr = '';

  // Añadir dígitos aleatorios a la cadena, asegurándose de que el primer dígito no sea 0
  numeroStr += random.nextInt(9).toString(); // El primer dígito no puede ser 0
  for (int i = 1; i < longitud; i++) {
    numeroStr +=
        random.nextInt(10).toString(); // Los siguientes dígitos pueden ser 0-9
  }

  return BigInt.parse(numeroStr); // Convertir la cadena a BigInt y devolverlo
}

BigInt generarPalindromoBigInt(int longitud) {
  String palindromoStr = '';
  String bigIntRandom = generarBigIntAleatorio(longitud ~/ 2).toString();

  palindromoStr = bigIntRandom + bigIntRandom.split('').reversed.join('');

  return BigInt.parse(palindromoStr);
}

/**
 * Notas
 * Convertir el BigInt a String es una de las operaciones que mas consume recursos.
 * Si esta operación no se utilizara en isPalindrome e isPalindrome2, la funcion
 * isPalindrome2 seria significativamente mas rapida.
 */

File: Desafio #01 - Numero palindromo/PHP/numeroPalindromoTest.php
<?php
function is_palindrome($number)
{
    if ($number != strrev($number)) {
        return false;
    }
    return true;
}
function is_palindrome2($number)
{
    $digits = strlen($number);
    $half_digits = floor($digits / 2);
    $digit_array = str_split($number);
    
    for ($i = 0; $i < $half_digits; $i++) {
        if ($digit_array[$i] != $digit_array[$digits - $i - 1]) return false;
    }

    return true;
}
// Función para generar un string de dígitos aleatorios
function generarStringAleatorio($longitud) {
    $string = "";
    for ($i = 0; $i < $longitud; $i++) {
        // Generar un dígito aleatorio entre 0 y 9
        $digito = rand(0, 9);
        // Concatenar el dígito al string
        $string .= $digito;
    }
    return $string;
}

// Generar un string de 100 dígitos aleatorios
$stringAleatorio = generarStringAleatorio(10000000);

// Imprimir el string generado
//echo $stringAleatorio;
echo "\n";

$inicioFuncionUno = microtime(true);
is_palindrome($stringAleatorio);
$finFuncionUno = microtime(true);
$tiempoFuncionUno = $finFuncionUno - $inicioFuncionUno;

// Medir el tiempo de ejecución de la función 2
$inicioFuncionDos = microtime(true);
is_palindrome2($stringAleatorio);
$finFuncionDos = microtime(true);
$tiempoFuncionDos = $finFuncionDos - $inicioFuncionDos;

// Comparar los tiempos de ejecución
echo "Tiempo de ejecución de la Función 1: " . $tiempoFuncionUno . " segundos\n";
echo "Tiempo de ejecución de la Función 2: " . $tiempoFuncionDos . " segundos\n";

if ($tiempoFuncionUno < $tiempoFuncionDos) {
    echo "La Función 1 es más rápida.\n";
} elseif ($tiempoFuncionUno > $tiempoFuncionDos) {
    echo "La Función 2 es más rápida.\n";
} else {
    echo "Ambas funciones tienen el mismo rendimiento.\n";
}

// Medir el consumo de memoria de la función 1
$inicioMemoriaFuncionUno = memory_get_usage();
is_palindrome($stringAleatorio);
$finMemoriaFuncionUno = memory_get_usage();
$consumoMemoriaFuncionUno = $finMemoriaFuncionUno - $inicioMemoriaFuncionUno;

// Medir el consumo de memoria de la función 2
$inicioMemoriaFuncionDos = memory_get_usage();
is_palindrome2($stringAleatorio);
$finMemoriaFuncionDos = memory_get_usage();
$consumoMemoriaFuncionDos = $finMemoriaFuncionDos - $inicioMemoriaFuncionDos;

// Comparar el consumo de memoria
echo "Consumo de memoria de la Función 1: " . $consumoMemoriaFuncionUno . " bytes\n";
echo "Consumo de memoria de la Función 2: " . $consumoMemoriaFuncionDos . " bytes\n";

if ($consumoMemoriaFuncionUno < $consumoMemoriaFuncionDos) {
    echo "La Función 1 consume menos memoria.\n";
} elseif ($consumoMemoriaFuncionUno > $consumoMemoriaFuncionDos) {
    echo "La Función 2 consume menos memoria.\n";
} else {
    echo "Ambas funciones consumen la misma cantidad de memoria.\n";
}